test/shared/src/main/scala/zio/test/TestRunner.scala

```diff
@@ -71,17 +72,27 @@ case class TestRunner[L, -T, E, S](
     spec: Spec[L, T],
     testLogger: TestLogger = defaultTestLogger,
     clock: Clock = Clock.Live
-  ): Exit[Nothing, ExecutedSpec[L, E, S]] =
-    Runtime((), platform).unsafeRunSync(run(spec).provide(buildEnv(testLogger, clock)))
+  ): Exit[Nothing, ExecutionResult[L, E, S]] =
+    buildRuntime(testLogger, clock).unsafeRunSync(run(spec))
 
   /**
    * Creates a copy of this runner replacing the reporter.
    */
   final def withReporter[L1 >: L, E1 >: E, S1 >: S](reporter: TestReporter[L1, E1, S1]) =
     copy(reporter = reporter)
 
-  private def buildEnv(loggerSvc: TestLogger, clockSvc: Clock): TestLogger with Clock = new TestLogger with Clock {
-    override def testLogger: TestLogger.Service = loggerSvc.testLogger
-    override val clock: Clock.Service[Any]      = clockSvc.clock
-  }
+  private[test] def buildRuntime(
+    loggerSvc: TestLogger = defaultTestLogger,
+    clockSvc: Clock = Clock.Live
+  ): Runtime[TestLogger with Clock] =
+    Runtime(buildEnv(loggerSvc, clockSvc), platform)
+
+  private def buildEnv(loggerSvc: TestLogger, clockSvc: Clock): TestLogger with Clock =
+    new TestLogger with Clock {
+      override def testLogger: TestLogger.Service = loggerSvc.testLogger
+      override val clock: Clock.Service[Any]      = clockSvc.clock
+    }
+}
+object TestRunner {
+  case class ExecutionResult[L, E, S](spec: ExecutedSpec[L, E, S], summary: String)
```

Comment:
Can we eliminate this and make it solely the responsibility of the `DefaultTestReporter` and the SBT test runner to filter for failing tests, convert them into string representations, aggregate them, and display them? I think this would give us a cleaner separation between "data" and "presentation of data". An `ExecutedSpec` should, I believe, have all of the underlying data on which tests passed and failed and the associated labels. Then it is up to the client to interpret those results however they want, including creating and displaying a summary like this.

The summary is very useful and we definitely want to include it in the default way we report test results but I think we want to avoid baking it into the return type for running tests to preserve flexibility for other clients to interpret the test results in different ways.

Reply:
Of course. This makes a lot of sense. I should commit this change during this weekend.